mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp

```diff
@@ -239,94 +459,255 @@ static LogicalResult checkUniqueReader(Operation *op, Operation *reader,
       }
     }
   }
-  if (nonViewReaders.size() != 1) {
+#ifndef NDEBUG
+  LLVM_DEBUG(llvm::dbgs() << "Non-view readers:\n");
+  for (Operation *reader : nonViewReaders)
+    LLVM_DEBUG(llvm::dbgs() << *reader << "\n");
+#endif
+  if (nonViewReaders.size() != 1)
     isUnique = false;
-  }
-  isUnique = (nonViewReaders[0] == reader);
+  else
+    isUnique = nonViewReaders[0] == reader;
   return success();
 }
 
-static ThreadwiseWriteAllOp traceToThreadwiseWrite(
-    Value inp, SmallVectorImpl<TransformMapAttr> &inpToTWWriteAllViews) {
+static Operation *
+traceToWriter(Value startVal,
+              SmallVectorImpl<TransformMapAttr> &writerToStartValViews) {
   // 1. Validate that the only uses of the linalg.generic input are the one
   // generic and a copy operation or transform.
-  ThreadwiseWriteAllOp result;
-  for (Operation *use : inp.getUsers()) {
-    SmallVector<Operation *, 4> nonViewUsers;
-    traceToNonViewUsers(use, nonViewUsers);
-    for (Operation *nonViewUse : nonViewUsers) {
-      if (isa<memref::DeallocOp>(nonViewUse)) {
-        // ignore
-        continue;
+  Operation *result = nullptr;
+  Operation *startValDef = startVal.getDefiningOp();
+  if (!startValDef)
+    return nullptr;
+  bool unique = true;
+  auto setResult = [&](Value val, Operation *theResult) {
```

Comment:
This might read better as a seperate function and with a description. 
I dont think there aren;t many local variables you need to capture to warrant a lambda here. Anything Im mising ?

Reply:
I'm capturing unique and result here.